bcipy/language/model/causal.py

```diff
@@ -0,0 +1,340 @@
+
+                    # Create a list of token indexes that are a prefix of target text
+                    # We go over all the integer IDs in the vocab and extra_vocab lists
+                    for i in itertools.chain(vocab, extra_vocab):
+                        hypo_str = sequence_text + self.index_to_word_lower[i]
+                        hypo_seq = batch_sequences[j].copy()
+                        hypo_seq += i,
+
+                        # Add the log prob of this token to the previous running total
+                        # For some reason the float cast makes it run faster
+                        likelihood = batch_likelihoods[j] + float(log_probs[j][i])
+                        # If we have extended to a space following the context, then that hypothesis gets to be done
+                        # This takes a lot longer that just requiring extending beyond existing context
+                        # Just require hypotheses to extend beyond the existing typed context
+                        if len(hypo_str) > len(context):
+                            ch = hypo_str[target_pos]
+
+                            # Create an empty list if we haven't seen this character before
+                            if ch not in char_to_log_probs:
+                                char_to_log_probs[ch] = []
+                            char_to_log_probs[ch] += likelihood,
+
+                        else:
+                            hypo = (likelihood, hypo_seq, hypo_str)
+                            if len(valid) < self.beam_width:
+                                heapq.heappush(valid, hypo)
+                            else:
+                                heapq.heappushpop(valid, hypo)
+
+        # Parallel array to symbol_set for storing the marginals
+        char_probs = []
+        for ch in self.symbol_set_lower:
+            # Convert space to the underscore used in BciPy
+            if ch == SPACE_CHAR:
+                target_ch = ' '
+            else:
+                target_ch = ch
+
+            # Handle cases when symbols are never seen
+            if target_ch in char_to_log_probs:
+                char_probs += logsumexp(char_to_log_probs[target_ch]),
+            else:
+                char_probs += float("-inf"),
+        # Normalize to a distribution that sums to 1
+        char_probs = softmax(char_probs)
+
+        next_char_pred = Counter()
+
+        for i, ch in enumerate(self.symbol_set_lower):
+            if ch is SPACE_CHAR:
+                next_char_pred[ch] = char_probs[i]
+            else:
+                next_char_pred[ch.upper()] = char_probs[i]
+
+        next_char_pred[BACKSPACE_CHAR] = 0.0
+
+        # width = 120
+        # print("             " + "_" * width)
+        # for i in range(len(self.symbol_set_lower)):
+        #     print(f"{self.symbol_set_lower[i]} = {char_probs[i]:4.2e} " + "*" * int(char_probs[i] * width))
+        # print("             " + "_" * width)
```

Comment:
We try to delete unused code to keep things tidy. It's always in version control if we need it back.